raft/src/main/java/org/apache/kafka/raft/FollowerState.java

```diff
@@ -36,8 +37,8 @@ public FollowerState(int epoch, Set<Integer> voters) {
     }
 
     @Override
-    public OptionalLong highWatermark() {
-        return highWatermark;
+    public Optional<LogOffsetMetadata> highWatermark() {
+        return highWatermark.isPresent() ? Optional.of(new LogOffsetMetadata(highWatermark.getAsLong())) : Optional.empty();
```

Comment:
Leaving the comment here for lack of an alternative. Let's leave this out of the scope of this PR, but I am wondering if we can get any benefit from this caching in the state machine, which is also sending fetches to the log. I think you one idea you mentioned is leaving the records in memory so that we could apply them to the state machine. I think fetches from the state machine will tend to be aligned with one of the previous high watermarks, so maybe a simple optimization like remembering the previous high watermark metadata would help.

Reply:
Yeah I do have another PR for caching the records directly, but I agree with you that most cases since we are reading from page cache anyways this may not benefit a lot, so I'm currently putting it on-hold, besides it is rather complicated (https://github.com/confluentinc/kafka/pull/344).

On the other hand though, since the hw caching only benefits the leader not the follower, I think we may still need to optimize for followers eventually.

Regarding your idea of caching the previous hw on follower, I've also thought about that while working on this PR, and here are a quick dump of my thoughts: we only apply to state machine when the follower's received fetch response indicate hw advanced, so I think the likelihood that a previously cached hw would be used in `updateFollowerHighWatermark` again may be since that means the hw was advanced beyond the follower's log-end offset, so previous `updateFollowerHighWatermark` did not apply to the hw, and then in the next fetch we may call `updateFollowerHighWatermark` again with the same hw.

Caching the previously appended info (that has the starting offset metadata) on the follower may be also useful since we know hw would only advance at the pace of batches that are aligned with fetch boundaries, but I'm not sure what would the "hit-ratio" be like either: since the hw advanced based on majority, the newly advanced hw may sits in the middle of a follower's previously fetched response if that spans over multiple batches e.g..

I think I can go ahead and create a JIRA summarizing those thoughts.

Reply:
Another idea I am considering is pulling the state machine out of `KafkaRaftClient` entirely. Rather than having the client pushing updates into the state machine and blocking the Raft IO thread, we can let the state machine be run in a separate thread and pull new data itself. If we did this, then optimizing the read overhead might not be as critical.

Reply:
I saw https://confluentinc.atlassian.net/browse/KC-578 already filed for this idea. I've added https://confluentinc.atlassian.net/browse/KC-650 for the other one.